Export frontmatter attributes

lostfictions opened the pull request:
I noticed that frontmatter from the markdown is being extracted, but discarded. This PR fixes that!

To accomplish this, I've upgraded `html-loader` to the latest version, so that we can pass in the [`esModule`](https://github.com/webpack-contrib/html-loader#esmodule) option. Without this, `html-loader` returns something like `module.exports='<html>...'`, which makes our job a lot more difficult.

With ES Modules, we can continue to default export the HTML, but also append named exports for each attribute. (I've opted for individual named exports here assuming this is better for tree-shakeability, but we could also, say, export a single value named `attributes` containing all the frontmatter values.)

As a bonus, this loader also can now accept a `htmlLoaderOptions` option to forward to `html-loader`. I've added all this to the readme.

Your lint configuration also didn't seem to be working; I've fixed that.

lostfictions commented:
As a side note, I at first wasn't sure why this module bundles `html-loader` rather than just suggesting users chain it themselves -- but with frontmatter exports the former approach makes a bit more sense to me. I'm not sure how one would achieve this with a separate `html-loader` definition (but I also haven't done a ton of loader writing).

skipjack commented:
> As a side note, I at first wasn't sure why this module bundles `html-loader` rather than just suggesting users chain it themselves -- but with frontmatter exports the former approach makes a bit more sense to me. I'm not sure how one would achieve this with a separate `html-loader` definition (but I also haven't done a ton of loader writing).

I actually had a note here that we should do exactly what you're describing:

https://github.com/skipjack/remark-loader/blob/master/src/index.js#L15-L16

Added specifically for image handling and it felt janky to me at the time too. I think if we can factor it out safely, that is the "correct" approach. I will take a look at this PR in more detail tonight (EST).

lostfictions commented:
> (I've opted for individual named exports here assuming this is better for tree-shakeability, but we could also, say, export a single value named attributes containing all the frontmatter values.)

Having used my fork in a project a bit more, I've realized just having a single `attributes` object export is probably a better approach here -- varying named exports make it impossible to have, say, TypeScript definitions for markdown imports. I'll push a change to this shortly.

lostfictions commented:
Okay, I've updated this PR so that frontmatter is assigned to an object exported as `attributes`. I've also added tests to this repo -- sorry if that means the PR is a bit big now.

skipjack commented:
First off, this is great thanks for your work! Given how much is here, I don't have time to go through all this tonight but I will by the weekend if not sooner. Here's some initial thoughts...

- I'd like to see if we can get @evilebottnawi's take on it all as I'm using this loader a bit less now.
- To your point, submitting the test suite and html loader options into separate prs would have been good.
- At this point, it's alright but please don't add anything else to this branch for now.
- You deleted the `html-loader` todo but I'm still thinking we should factor that out of this package and recommend chaining (doesn't have to happen in this pr, but we should keep the todo).

skipjack commented on src/index.js:
```diff
@@ -4,19 +4,34 @@ const parse = require('./parse.js')
+
     parse(content, options)
-        // @todo we should probably just intercept images in the tree
-        // or recommend that the `html-loader` be chained
```
As noted in my previous comment, I would keep the `@todo` unless it's really impossible or impractical for us to recommend chaining here.

lostfictions replied:
Agreed that it would be ideal to just leave this to webpack. As it stands, I couldn't figure out how to make this work, because this loader needs to intervene both before _and_ after `html-loader`. Currently the loader passes its output to `html-loader`, receives the result, and then appends the frontmatter export to that result. Would love to hear if there's a smarter way to do this!

skipjack replied:
Indeed, which may have been why I did it that way in the first place. I'll do a little research on this front and get back to you.

In the short term I would keep the todo and then maybe the best next steps would be to merge this as is and create an issue where we discuss a more robust refactor that allows for chaining and supports frontmatter via `plugins` rather than the current approach.

skipjack commented on package.json:
```diff
@@ -19,19 +19,42 @@
     "lint": "eslint .",
-    "release": "standard-version"
+    "release": "standard-version",
+    "test": "jest"
```
I ran the test suite and everything passed, however it is throwing this warning at the end:

![image](https://user-images.githubusercontent.com/8988697/87095658-20465680-c20f-11ea-9a8c-63ae1f52c86d.png)

lostfictions replied:
Not sure about this -- the tests are somewhat cribbed from `html-loader`'s, so maybe @evilebottnawi has more insight on what's happening. (Actually, it might be ideal if there were a shared loader testing library for `webpack-contrib` repos....?)

lostfictions commented on README.md:
````diff
@@ -36,19 +36,51 @@ module.exports = {
 `plugins` list.
 
-> This loader makes use of the [`html-loader`][5] under the hood.
+Frontmatter values are assigned to an object exported as `attributes`. For
+example, if you have this markdown file:
+
+```md
+---
+title: a clever remark
+---
+hello world!
+```
+
+You can use it like this:
+
+```js
+import contents, { attributes } from './test.md'
+
+console.log(contents) // logs "hello world!"
+console.log(attributes) // logs "{ title: 'a clever remark' }"
+```
````
I was going by what's already in the library, which already parses frontmatter via the `front-matter` package. Agreed that it could be better to leave that to userspace as a remark plugin! (That said, I think the use of frontmatter in markdown files is so common that it could be useful to include it as an opinionated default, similar to how `remark-html` is included now.)

skipjack replied:
🤦‍♂️ wow that shows how long it's been since I touched this. Completely forgot that, I must've added that as a quick hack before I knew about the plug-in. Would definitely be good to get rid of that and somehow this via the plug-in instead. Just to clarify, there already is a frontmatter plugin for remark, I linked it above.

lostfictions commented on package.json:
```diff
@@ -19,19 +19,42 @@
+    "jest": "^26.1.0",
+    "memfs": "^3.2.0",
+    "standard-version": "^4.2.0",
+    "webpack": "^4.43.0"
```
Sounds good -- I didn't upgrade any packages other than the ones needed for this PR.

skipjack replied:
Yeah those additions may have vulnerabilities though, maybe try an `npm audit fix` on this branch and see it knocks some out?

evilebottnawi commented:
Why do you need this? I think it is out of scope this loader

evilebottnawi commented:
We did full refactor - using loader inside loader is bad idea, we split them, so you need to install `html-loader` separate, also you can any own loader, example https://github.com/webpack-contrib/remark-loader#markdown-to-html, you can even do markdown to markdown and other nice stuff

lostfictions commented:
@evilebottnawi Hi, it's still not clear to me with this refactor whether it's possible to take markdown with frontmatter attributes, and import it as an object that provides both the markdown converted to HTML **as well as** the frontmatter attributes. This is a very common use case -- in fact, I'd argue that _by far the most common use case_ of markdown with frontmatter is to convert it to HTML _and_ an object of frontmatter attributes. That is the use case this PR was meant to address. Generally when one supplies frontmatter in a markdown file, it is because they need access to this metadata -- in the general case I don't think people want to just throw away the frontmatter.

I understand that when webpack took over ownership of this repo, they decided to ignore all pending PRs and simply rewrite the package, but it's pretty disappointing to me that they wouldn't consider IMO what's an actual significant use case. As far as I can tell, given the way the package exist now, if I wanted to handle this use case of HTML and frontmatter, I'd need to fork `html-loader` or write an entirely new loader that bundles `html-loader` and defers to it anyway.

Am I incorrect? Is there a simple way using this loader to convert a markdown file with frontmatter such that I can import both HTML output and frontmatter attributes?

evilebottnawi commented:
@lostfictions 

> I understand that when webpack took over ownership of this repo, they decided to ignore all pending PRs and simply rewrite the package, but it's pretty disappointing to me that they wouldn't consider IMO what's an actual significant use case.

It's a shame to hear, because I asked you before the release, but you showed no activity, so I figured that we can figure it out later.

I am a bit confused about what you want to solve exporting attributes? I would say more - it can potentially break some cases. If you presented short example/repo with readme what you are trying to achieve, we would be able to find a better and more convenient solution for you, without this context, I cannot just take and start export something, tomorrow somebody can ask export full ast, then something else, and you bundle will be big and slow

lostfictions commented:
> It's a shame to hear, because I asked you before the release, but you showed no activity, so I figured that we can figure it out later.
I'm sorry if that's the case, but I'm not sure it is? I think both @skipjack and I pinged you in this PR a month ago, but it looks you only replied to this thread about two hours before the release was cut. I'm sorry if I wasn't able to jump on the issue to reply that quickly.

> I am a bit confused about what you want to solve exporting attributes? I would say more - it can potentially break some cases. If you presented short example/repo with readme what you are trying to achieve, we would be able to find a better and more convenient solution for you, without this context, I cannot just take and start export something, tomorrow somebody can ask export full ast, then something else, and you bundle will be big and slow

That seems like a kind of absurd counterexample? I don't think frontmatter attributes are irrelevant metadata or some implementation detail like an AST. For example, most of the popular tooling that consumes markdown files explicitly makes use of frontmatter. [Jekyll](https://jekyllrb.com/docs/front-matter/) (used by default for Github Pages) employs it to determine layout template as well as variables that can be interpolated into templates; [Gatsby](https://www.gatsbyjs.org/packages/gatsby-transformer-remark/?=#how-to-query) allows querying it; [Hugo](https://gohugo.io/content-management/front-matter/) makes extensive use of it as well. Webpack's own website, webpack.js.org, [makes extensive use of frontmatter attributes](https://github.com/webpack/webpack.js.org/tree/master/src/content) in its markdown files. I feel it's very likely that if people start to use this loader more extensively, this issue will be raised again, because in my opinion it's a significant one.

I don't have a full public repo with an example of this, but as a very short sketch here's how usage of frontmatter attributes as proposed by this PR could work:

```jsx
// ...

const { default: content, attributes } = await import(`content/${page}/${lang}.md`);

const { title, publishDate, published } = attributes;

// render a page statically using the given html body `content`, interpolating
// the title and publication date into the template, predicated on whether the
// `published` content attribute is true/false or undefined.
```

I can try extract a more extensive example (rendering to static HTML with React via Next.js, since that's the context I'm currently using it in) if that would be helpful or more illustrative of why frontmatter is valuable.

skipjack commented:
> I'm sorry if that's the case, but I'm not sure it is? I think both @skipjack and I pinged you in this PR a month ago, but it looks you only replied to this thread about two hours before the release was cut. I'm sorry if I wasn't able to jump on the issue to reply that quickly.

Hey @lostfictions, this is at least partially on me since I haven't had time for the repo and unlucky timing during the transition. You definitely got burned a little here, but I think we can still come to some sort of resolution for what you're proposing.

I agree the use case you were shooting for is common and what you outlined is clear. I think there's probably a good way to support it we just have to think of how to pass an additional expert through the next loader in the chain. @evilebottnawi @sokra @bebraw is that even possible?

@lostfictions I think the most logical next step is to create an issue with what you outlined above so we can brainstorm more there. Unfortunately this pr should probably be closed at this point but you can always keep your fork and install it through npm's GitHub url feature.

evilebottnawi commented:
@lostfictions remark loader do not export JS, you can look at source code, but we can `AST`/`frontmatter`/etc transfer to next loader (export created by html-loader)

sokra commented:
A loader can return multiple values. The first one is usually the content, that would be HTML here. The second one is usually a SourceMap, not sure if this is available here but we should reserve it for that at least. The 3rd one is usually an options object with more things. You could put an `attributes` object here. The html-loader could be extended to handle this.

It's also possible without html-loader modifications, but this would require an additional loader which adds the additional export after js generation. This is less UX friendly and more complex, so better go with the first approach.

skipjack commented:
Those details are exactly what we needed and resolve the discussion we started having on the initial changes in this pull request. @lostfictions what @sokra laid out above is exactly what we need...

> A loader can return multiple values. The first one is usually the content, that would be HTML here. The second one is usually a SourceMap, not sure if this is available here but we should reserve it for that at least. The 3rd one is usually an options object with more things. You could put an `attributes` object here. The `html-loader` could be extended to handle this.

I think the path forward is fairly obvious now. Let's create two issues...

- One in the [html-loader][1] to allow passing meta data like frontmatter as an additional export.
- One in the this repo to to allow users to somehow configure the frontmatter handover to the next loader.

The complete picture won't work until both changes are in place but they should be able to be addressed in parallel and then tied together.

Again sorry this PR didn't make it through, but once we flesh out the details in those issues my guess is that @evilebottnawi and the rest of the team would welcome more targeted PRs to address those two steps and provide support for what you described.

[1]: https://github.com/webpack-contrib/html-loader

lostfictions commented:
@skipjack Thanks for following up on this! And sorry for my frustration that this PR was passed over -- I totally get that it was leaning on the pre-rewrite hackish approach of this loader, and that modifying html-loader and this loader to support some kind of handover as you and Tobias outlined makes more sense.

I'm not sure I have the capacity to write up these issues (or take a pass at the necessary PRs) in the next little while -- for better or for worse my hackish fork is working well for the project I needed it for -- but if nobody else gets to it sooner I could maybe take a look in a few weeks.

skipjack commented:
No worries, let me know if you'd like any help creating them. Please cc me on them and reference this pr if you do end up creating them.

evilebottnawi commented:
@lostfictions I can help, it is not hard, please open a new issue, otherwise we will lost the issue
